start to-do back end , install first dependences for project   ,make …

mhmdwSamir opened the pull request:
…connection to mongodb , creat model for task

mhmdwSamir committed:
start to-do back end , install first dependences for project   ,make connection to mongodb , creat model for task

mhmdwSamir committed:
create simple  end point for add task  ,
create custom class [ Exception ] to handle errors and an object for http-status-code  to grap them in core folder .
create task router file to grap all task routes .

mhmdwSamir committed:
create controller folder and add task controller that contain add task functionality

mhmdwSamir committed:
create sub folder from core to db and isolate it from app.js starter file

mhmdwSamir committed:
delete db.js file after creating it in core folder.
add cors to enable requests from frontEnd

mhmdwSamir committed:
create simple end point to get all tasks and run it

mhmdwSamir committed:
fix getAllTasks end point to be suitable with some case ,
adding http status code 204

mhmdwSamir committed:
add deleteTaskById function in task controller and handle some cases of errors in it and stuck in others .

mhmdwSamir committed:
- make new function that get  active tasks only .
- make new function that get completed  tasks only .
- add new status code in http_status_code.js | not found:404.
- add routes for new 2 functions  one for get active tasks and the other for gget complete tasks

ahmedradwan94 commented on controller/task.controller.js:
```diff
@@ -0,0 +1,98 @@
+      });
+      res.send(tasks);
+    } catch (error) {
+      res.status(404).send(error);
```
Error handling is not correct.
your implementations says that: `if there's any type of error in the code in try {} block server will respond with HTTP status code 404 which means NOT_FOUND err`.
And actually it may occurs any type of error that's not a `NOT_FOUND`.

ahmedradwan94 commented on controller/task.controller.js:
```diff
@@ -0,0 +1,98 @@
+      let { content, completed } = req.body;
+      // validate it
+      if (!content)
+        throw new Exception(
```
Great validation, good job man.

ahmedradwan94 commented on controller/task.controller.js:
```diff
@@ -0,0 +1,98 @@
+          "BadxxR"
+        );
+      // create task with req body and save it in db
+      let task = new Task({ content, completed });
```
In create task we must not let the `client-app` to create new task with `{ completed: true }`.
so I'll expect from client-app to send me only the content and I'll create the task with status `{ completed: false }`.

ahmedradwan94 commented on controller/task.controller.js:
```diff
@@ -0,0 +1,98 @@
+      console.log(_id);
+      // handle case if no id was provided
+      if (!_id)
+        throw new Exception(
```
Nice validation

ahmedradwan94 commented on controller/task.controller.js:
```diff
@@ -0,0 +1,98 @@
+        );
+      let task = await Task.findById(_id);
+      if (!task)
+        throw new Exception(
```
Nice validation also.

ahmedradwan94 commented on core/db/db.js:
```diff
@@ -0,0 +1,18 @@
+const mongoose = require("mongoose");
```
Please use **single** quote instead of **double** quotes.

ahmedradwan94 commented on core/db/db.js:
```diff
@@ -0,0 +1,18 @@
+const mongoose = require("mongoose");
+const conectDb = async () => {
+  await mongoose
+    .connect("mongodb://localhost/todoapp", {
```
You must get the mongoose `URI` from `process.env` environment variables.
so if you will to deploy this application in a host you should provide the MongoDB URI in environment variables.
example:
`await mongoose.connect(process.env.MONGO_URI);`

ahmedradwan94 reviewed (changes requested):
Great job @mhmdwSamir, There's a few `requested changes`.

mhmdwSamir committed:
- create new route for user.
- add sign up function to user  to can sign up .
- add ` deleteCompletedTasks ` function to delete all completed tasks.
- update app.js file with new changes .
- install dotenv packahe to holde all varaible and use it

mhmdwSamir committed:
fix bug in using .env file
- replace static numbers from app.js file and sign up function from variable from .env file
- remove uncessary code from sign up function

a7mdRadwan commented on app.js:
```diff
@@ -0,0 +1,22 @@
+// user route
+app.use("/api", userRouter);
+
+app.listen(parseInt(process.env.PORT), () => {
```
the next time add a default value to protect your application from failure.
so if this env var not defined your application will work.
ex: `process.env.PORT != null ? parseInt(process.env.PORT) : 3000`.

a7mdRadwan commented on app.js:
```diff
@@ -0,0 +1,22 @@
+
+/* routes */
+// task route
+app.use("/api", taskRouter);
```
it should be `/api/task`.

a7mdRadwan commented on app.js:
```diff
@@ -0,0 +1,22 @@
+// task route
+app.use("/api", taskRouter);
+// user route
+app.use("/api", userRouter);
```
it should be `/auth/user`

a7mdRadwan reviewed (changes requested):
@mhmdwSamir  Please resolve the comments that you solved in your code.

mhmdwSamir committed:
create new end point to make user login .
- use jwt to creat token for user .
- update user.router.js file to this new method

mhmdwSamir committed:
achieve login story in peace
- create separte function require-auth that check if the user is already login in middleware folder in core folder
